File: synapsys-benchmarks/src/jmh/java/io/eigr/synapsys/benchmarks/SimpleActor.java
package io.eigr.synapsys.benchmarks;

import org.jetbrains.annotations.NotNull;

import io.eigr.synapsys.core.actor.Actor;
import io.eigr.synapsys.core.actor.Context;
import kotlin.Pair;

public class SimpleActor extends Actor<Integer, Message, String> {

    public SimpleActor(@NotNull String id, Integer initialState) {
        super(id, initialState);
    }

    @Override
    public @NotNull Pair<Context<Integer>, String> onReceive(@NotNull Message message, @NotNull Context<Integer> ctx) {
        return new Pair<>(
                ctx.withState(ctx.getState() + 1),
                "Processed: " + message.text() + " with new state: " + ctx.getState()
        );
    }
}

File: synapsys-benchmarks/src/jmh/kotlin/io/eigr/synapsys/benchmarks/SimpleActor.kt
package io.eigr.synapsys.benchmarks

import io.eigr.synapsys.core.actor.Actor
import io.eigr.synapsys.core.actor.Context
import org.slf4j.LoggerFactory

class SimpleActor(id: String?, initialState: Int?) : Actor<Int, Message, String>(id, initialState) {
    private val log = LoggerFactory.getLogger(SimpleActor::class.java)

    override fun onReceive(message: Message, ctx: Context<Int>): Pair<Context<Int>, String> {
        log.info(
            "Received message on Actor {}: {} with state: {} on thread {}",
            id,
            message,
            ctx.state,
            Thread.currentThread()
        )

        val newCtx = ctx.withState(ctx.state!! + 1)
        return ctx to "Processed: ${message.text} with new state: ${newCtx.state}"
    }
}